### DISCUSSION TITLE
WAV audio file contains TWO EMPTY CHANNELS (and apparently NON-RECOVERABLE hard-clipping, even though its 32-bit FLOAT)

### DISCUSSION THREAD

**Discussion by RobertMarkEnger** [gopro/labs#742] [Feature Requests]
I am using a GoPro 11, with media mod.  I have the LABS software installed, version 02.12.70.
I am recording raw WAV files, set to minimal (no processing).
Only two microphones are in use when the media-mod is installed.
BUT, the GoPro creates  WAV files with FOUR AUDIO CHANNELS.   TWO CHANNELS ARE EMPTY, but take up space on the mem card.
Maybe, when the MediaMod is installed, the system should only create TWO CHANNEL wav files?  Save space on the memory card!

ALSO, you will note that there are CLIPPING EVENTS.  (red)
Supposedly, with 32-bit FLOAT recording, it is still possible to recover values with level greater than 1.0.
The instructions suggest to use a DAW (such as Audacity) and to use the "Amplify" software effect, with a NEGATIVE gain value.
This processes the audio file, reducing levels.  With enough negative-gain, the values will all be brought down below 1.0.

I did this with the TOP trace.  (see 2nd picture below).  
You will note that while the RED is gone, the signal is never-the-less **FLAT TOP HARD CLIPPED**.
Even though GoPro is creating 32-bit FLOAT files, it is NOT creating audio where clipping events can be recovered.

IT WOULD BE NICE TO HAVE RECOVERABLE CLIPPING EVENTS, AS IMPLIED BY THE USE OF 32-BIT FLOAT RECORDING FORMAT!

ZoomCorp and others utilize TWO  ADC-chips and associated control circuitry to ensure that non-recoverable hard-clipping does not (easily) occur.
It would be nice if GoPro provided some hardware-enhancement that addressed this issue head-on.

As a work-around, it would be nice if LABS provided some way for us to attenuate the analog signal going into the ADC chip.
With some trial-and-error preparation, we could find some attenuation setting that was appropriate for the usage scenario.
(In my case, this is putting cameras into demolition derby cars during the race.)
The wav files can sound VERY NICE (compared to the compressed-pabulum AAC file).  Avoiding clip events makes the WAV files more usable (and thus more valuable)!

My current plans are to deploy a dedicated (time-code equipped) audio recorder into the car.  This is on top of TWO GoPro cameras.  It is a shame not to be able to leverage the four WAV audio channels of the two GoPro cameras.  
But they have no TimeCode, and they suffer non-recoverable hard-clipping events.

![GoProLABS_wav_file_4-chans](https://github.com/gopro/labs/assets/157267205/6b5d0c28-9928-4050-a34a-68c5591eddaf)
![GoProLABS_wav_file_32float_hard-clipped](https://github.com/gopro/labs/assets/157267205/be7ffbee-be4a-4850-868d-5430e923b8cb)

**Comment by dnewman-gpsw** [comment]
Can you share a WAV file that clips, as I can't get an audio signal load enough to cause this? This waveform is me yelling 1" away from MediaMod Front mic (measuring in at ~125dB SPL -- yelling inside my car).  
![image](https://github.com/gopro/labs/assets/10856084/547c02f5-cb81-49f2-aab8-e16e49087c51)

You are correct 32-bit float as infinite headroom, however hardware limitations of microphones and AGC configurations, will set an upper boundary.  I don't know what SPL MediaMod can handle, can't find that spec, but many vendors microphones are limited to 135dB maximum SPL.  135dB is a lot.

**Reply by RobertMarkEnger** [reply]
hello David:
Thank you for reviewing my post.  The loud events appear to occur when ANOTHER CAR (with particularly loud exhaust) passes close to the one we had the GoPro installed in (some discussion at the bottom).

It is possible that the mic elements themselves could clip.  Or that the 1st-stage FET pre-amp associated with the mic element overloads.  And, of course the analog input electronics of the GoPro could overload.  Finally, the ADC could be driven to 0dBfs (all-1s clip).

I think ZoomCorp, Sound Devices, Tascam, and others with 32-bit Float recorders all include two ADC chips (and probably, although unstated, some degree of parallelized analog input electronics, to ensure clipping in the analog input stages does not corrupt the loud passages of audio).  They configure one of the ADCs to encounter all-1s clip at a substantially higher input level than the other (say 30dB).   Loud-passage values from this second ADC are used to generate 32-bit FLOAT values above 1.0.   (TentacleSync makes an audio recorder, and they too claim to support 32-bit Float, but when pushed for details, they will not go on the record that they have implemented dual ADC.)  

32-bit FLOAT is not really giving us infinite headroom.  It is just a signal encoding protocol.   It is the presence of enhanced ADC systems (e.g. DUAL ADC chips, and careful attention to analog overload levels) that is increasing the range of usable input signal levels. 

Perhaps the LABS folks can work with GoPro hardware group to recommend improvements to the audio hardware chain for some future version of the product?

[DITTO FOR TIME CODE.   I HOPE GOPRO WILL SWITCH TO A HIGHER GRADE OSCILLATOR (**txco**) THAT WILL HOLD SYNC LONGER THAN A HALF AN HOUR AFTER JAM.] 

In the mean time, if new software could allow us to reduce the input level to the ADC, we could at least try that in the field and see if any improvement can be obtained.  (if it is just the ADC going all-1s, then we win, assuming missing time-code is eventually fixed.  If indeed, some analog section ahead of the ADC is hitting the wall, then no hope until a hardware redesign.)

I considered using external MICs and signal-attenuation (pads) to reduce the signal level before it enters the GoPro.
But I am still faced with lack of time-code support for the WAV files.  
This led me to go out and purchase dedicated audio gear (with long-term stable TXCO time-code).  
Seems a shame, with two modern GoPros in the car already. 

Apparently, GitHub won't accept WAV files for direct upload...

**I uploaded the file to Google Drive.**  
https://drive.google.com/file/d/1q4KOaHSKnS1YmWHpFSOiPo4AP0RZwsMY/view?usp=sharing 

(I hope GDrive doesn't try to recompress it, or otherwise modify it.  Hash of original file is below.)

`SHA256            04771C08ADCF2BED283DA28406636F453A160F6BEA6C3026C2463C03FF2645EA   `


Thank you again for reviewing my notes!
(info on the audio file and demo derby in general is below)

Some of these folks operate cars with straight-pipe exhaust.  Makes a lot of noise and the audiences love that.  
The audio is from an event using smaller, mostly un-modified cars.  (no gonzo reinforcing of the bumpers, etc)
In the larger class of competition, a couple of the cars have direct-pipe exhausts that exit the engine compartment vertically, straight up.  They apparently inject fuel into the pipes so that flames shoot out of the top of the pipes.  Very popular with the crowds.

I didn't know much about demolition derbies.  I am assisting a videographer who is creating a documentary about women drivers in the derbys.  I am a retired electrical engineer acting as location-sound person, and in-car B-cam tech.  My wife runs wide-shot B-cam with an FX3.  The videographer we are helping has fancier pro gear.

**Reply by dnewman-gpsw** [reply]
The ACG will already be at zero gain, so I expect the Microphone and circuit design are hitting there limits at around 130dB.  There is a metadata track in the MP4 that monitors the ACG, with average and peak input levels at 10Hz.  It reports in dbFS (Full Scale.)  My ~125db SPL yelling had measured -4dbFS peak audio level. So my own crude test would have an approximate 130dB limit to inputs on MediaMod hardware.  It is likely impractical to redesign hardware for such rare situations, MediaMod being mainly designed for Vlogger and Studio uses (my guess.) Instead you can reduce the problem at the source, by attenuating the audio input to the microphone. Add packaging tape to the microphone under the wind muff, leaving only the front element exposed (otherwise too muffled), in my crude testing, this is 5-6dB more headroom.  Labs is about hacks after all. ;) 

![image](https://github.com/gopro/labs/assets/10856084/3ca115af-248e-483f-96c1-a94d51912e9e)

If my MediaMod looks a bit weird, is it just sawed-off, otherwise stock.
![image](https://github.com/gopro/labs/assets/10856084/1eb9d3a8-0b44-437b-9e47-99453f49c6d8)

**Reply by RobertMarkEnger** [reply]
Well, if they fix the TimeCode on wav files, I can go with external mics and attenuators.  (some mics even have switchable attenuation capability built-in).  Using the Pro Mic Adapter probably makes more sense, if we ever get to go that route.
But without timecode, I think editors will put a contract out on my life.  I'll stick with the TC-equipped recorder for now.

**Comment by moll** [comment]
Hey, did you ever find out what's with the two empty channels? I'm seeing the same thing and thought the 4 channels would hold the internal mics _and_ the external mic (attached to the media mod). I'm still on stock firmware, though, so perhaps that functionality starts working only on Labs.

**Comment by sjmarcy2** [comment]
FWIW Hero 13 WAV audio seems to have a consistent just over 1/10s delay compared to the standard audio when viewed on Audacity.   Easy to handle of course.   I prefer to use the level 2 processing on WAV with media mod fed separate lav mics.    Level 3 or standard has too much AGC for Motorsports use, for instance it boosts up a quiet engine Idle.   AGC seems to have different transient behavior, and sometimes can "click" more than the lesser processed options..especially in the first few seconds.   Overall very pleased, would be nice to sort out channels and BT mic(s), esp if BT audio has a wide bandwidth and stereo.    Thanks!
